### ISSUE TITLE
Tensor subclass boilerplate can be consolidated

### ISSUE THREAD

**Issue by HDCharles** [pytorch/ao#710]
A lot of code for tensor subclasses can likely be conslidated together into a base class that other classes can utilize

_get_to_kwargs:
https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py#L64
https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py#L276
(also needed for https://github.com/pytorch/ao/blob/main/torchao/quantization/autoquant.py#L40)

to:
https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py#L594
https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py#L290
https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py#L423
(also needed for https://github.com/pytorch/ao/blob/main/torchao/quantization/autoquant.py#L40)

_apply_fn_to_data:
https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py#L432

detach:
https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py#L444

a default __repr__ would be nice (it caused bugs in the past)

if these were implemented using __tensor_flatten__ and __tensor_unflatten__ e.g. https://github.com/pytorch/ao/blob/main/torchao/dtypes/affine_quantized_tensor.py#L412-L421
you could write a general mixin/parent class that handles the general form of those methods

**Comment by HDCharles** [comment]
@jerry39213gh thoughts?

**Comment by jerryzh168** [comment]
yes, this makes sense I think, we can put this https://github.com/pytorch/ao/blob/main/torchao/dtypes/utils.py. we can also attach https://github.com/pytorch/ao/blob/ac8ce4ceb548c264a2f508f7b6a1f413d8d4454c/torchao/quantization/linear_activation_quantized_tensor.py#L98-L100 by default

there could be some issues of child class modifying the dispatch table for parent class, if this becomes an issue, we can add `class` to the dispatch table as well in the future.
